packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts

```diff
@@ -951,50 +1157,80 @@ export class AutoScalingGroup extends AutoScalingGroupBase implements
     }
 
-    this.securityGroup = props.securityGroup || new ec2.SecurityGroup(this, 'InstanceSecurityGroup', {
-      vpc: props.vpc,
-      allowAllOutbound: props.allowAllOutbound !== false,
-    });
-    this.connections = new ec2.Connections({ securityGroups: [this.securityGroup] });
-    this.securityGroups.push(this.securityGroup);
-    Tags.of(this).add(NAME_TAG, this.node.path);
-
-    this.role = props.role || new iam.Role(this, 'InstanceRole', {
-      roleName: PhysicalName.GENERATE_IF_NEEDED,
-      assumedBy: new iam.ServicePrincipal('ec2.amazonaws.com'),
-    });
-
-    this.grantPrincipal = this.role;
-
     if (props.groupMetrics) {
       this.groupMetrics.push(...props.groupMetrics);
     }
 
-    const iamProfile = new iam.CfnInstanceProfile(this, 'InstanceProfile', {
-      roles: [this.role.roleName],
-    });
+    let launchConfig: CfnLaunchConfiguration | undefined = undefined;
+    if (props.launchTemplate || props.mixedInstancesPolicy) {
+      this.verifyNoLaunchConfigPropIsGiven(props);
 
-    // use delayed evaluation
-    const imageConfig = props.machineImage.getImage(this);
-    this.userData = props.userData ?? imageConfig.userData;
-    const userDataToken = Lazy.string({ produce: () => Fn.base64(this.userData.render()) });
-    const securityGroupsToken = Lazy.list({ produce: () => this.securityGroups.map(sg => sg.securityGroupId) });
-
-    const launchConfig = new CfnLaunchConfiguration(this, 'LaunchConfig', {
-      imageId: imageConfig.imageId,
-      keyName: props.keyName,
-      instanceType: props.instanceType.toString(),
-      instanceMonitoring: (props.instanceMonitoring !== undefined ? (props.instanceMonitoring === Monitoring.DETAILED) : undefined),
-      securityGroups: securityGroupsToken,
-      iamInstanceProfile: iamProfile.ref,
-      userData: userDataToken,
-      associatePublicIpAddress: props.associatePublicIpAddress,
-      spotPrice: props.spotPrice,
-      blockDeviceMappings: (props.blockDevices !== undefined ?
-        synthesizeBlockDeviceMappings(this, props.blockDevices) : undefined),
-    });
+      if (props.launchTemplate && props.mixedInstancesPolicy) {
+        throw new Error('Setting \'mixedInstancesPolicy\' must not be set when \'launchTemplate\' is set');
+      }
+
+      if (props.launchTemplate && props.launchTemplate instanceof ec2.LaunchTemplate) {
+        this.launchTemplate = props.launchTemplate;
+      }
 
-    launchConfig.node.addDependency(this.role);
+      if (props.mixedInstancesPolicy && props.mixedInstancesPolicy.launchTemplate instanceof ec2.LaunchTemplate) {
+        this.launchTemplate = props.mixedInstancesPolicy.launchTemplate;
+      }
+
+      this._role = this.launchTemplate?.role;
+      this.grantPrincipal = this._role || new iam.UnknownPrincipal({ resource: this });
+
+      this.osType = this.launchTemplate?.osType || ec2.OperatingSystemType.UNKNOWN;
+    } else {
```

Comment:
This `if else` statement (the `if` on line 1165, and the `else` on 1184) is far too big. Can we extract the code in each case to its own function? So we'd have:

```
    if (props.launchTemplate || props.mixedInstancesPolicy) {
      handleIfCase();
    } else {
      handleElseCase();
    }
```

(of course, with better-named functions)